platforms/common/px4_work_queue/WorkQueueManager.cpp

```diff
@@ -248,46 +253,113 @@ static void WorkQueueManagerRun()
 	}
 }
 
-int WorkQueueManagerStart()
+int
+WorkQueueManagerStart()
 {
-	int task_id = px4_task_spawn_cmd("wq:manager",
-					 SCHED_DEFAULT,
-					 PX4_WQ_HP_BASE,
-					 1200,
-					 (px4_main_t)&WorkQueueManagerRun,
-					 nullptr);
-
-	if (task_id < 0) {
-		PX4_ERR("task start failed (%i)", task_id);
-		return -errno;
+	if (_wq_manager_should_exit.load() && (_wq_manager_create_queue == nullptr)) {
+
+		_wq_manager_should_exit.store(false);
+
+		int task_id = px4_task_spawn_cmd("wq:manager",
+						 SCHED_DEFAULT,
+						 PX4_WQ_HP_BASE,
+						 1280,
+						 (px4_main_t)&WorkQueueManagerRun,
+						 nullptr);
+
+		if (task_id < 0) {
+			_wq_manager_should_exit.store(true);
+			PX4_ERR("task start failed (%i)", task_id);
+			return -errno;
+		}
+
+	} else {
+		PX4_WARN("already running");
+		return PX4_ERROR;
 	}
 
-	return 0;
+	return PX4_OK;
 }
 
-int WorkQueueManagerStop()
+int
+WorkQueueManagerStop()
 {
-	if (_wq_manager_wqs_list != nullptr) {
-		auto lg = _wq_manager_wqs_list->getLockGuard();
+	if (!_wq_manager_should_exit.load()) {
 
-		// ask all work queues (threads) to stop
-		// NOTE: not currently safe without all WorkItems stopping first
-		for (WorkQueue *wq : *_wq_manager_wqs_list) {
-			wq->request_stop();
+		// error can't shutdown until all WorkItems are removed/stopped
+		if ((_wq_manager_wqs_list != nullptr) && (_wq_manager_wqs_list->size() > 0)) {
+			PX4_ERR("can't shutdown with active WQs");
+			WorkQueueManagerStatus();
+			return PX4_ERROR;
+		}
+
+		// first ask all WQs to stop
+		if (_wq_manager_wqs_list != nullptr) {
+			{
+				auto lg = _wq_manager_wqs_list->getLockGuard();
+
+				// ask all work queues (threads) to stop
+				// NOTE: not currently safe without all WorkItems stopping first
+				for (WorkQueue *wq : *_wq_manager_wqs_list) {
+					wq->request_stop();
+				}
+			}
+
+			// wait until they're all stopped (empty list)
+			while (_wq_manager_wqs_list->size() > 0) {
```

Comment:
Shouldn't you still have the lock guard here?

Reply:
There's one in size(). https://github.com/PX4/Firmware/blob/f1339038ca74b4d6da75fade34f2eb48bce8f643/src/include/containers/BlockingList.hpp#L71-L75